src/HartreeParticleDSL/Particle_IR/symbols/symboltable.py

```diff
@@ -0,0 +1,168 @@
+        :raises IRGenerationError: if the symbol table already contains a symbol of this name.
+        '''
+        if not isinstance(new_symbol, Symbol):
+            raise IRGenerationError(f"Symbol {new_symbol} is not a symbol, but "
+                                f"{type(new_symbol)}.")
+    
+        if new_symbol.name in self._symbols:
+            raise IRGenerationError(f"Tried to add a new symbol {new_symbol.name} "
+                                    "but it was already present in the symbol table.")
+    
+        self._symbols[new_symbol.name] = new_symbol
+    
+    def new_symbol(self, name: str, datatype: DataType, symbol_type: type[Symbol]) ->  Symbol:
+        '''
+        Create a new symbol of symbol type `symbol_type`, with name `name` and datatype `datatype`.
+        The visibility will match the visibility of this symbol table.
+    
+        :param name: Name of the symbol. This name must be unique in this symbol table.
+        :type name: str
+        :param datatype: The data type of the symbol.
+        :type datatype: :py:class:`HartreeParticleDSL.Particle_IR.datatypes.datatype.DataType`
+        :param symbol_type: The Particle IR type of the symbol to create.
+        :type symbol_type: :py:type:`HartreeParticleDSL.Particle_IR.symbols.symbol.Symbol`
+    
+        :raises TypeError: if the type_symbol argument is not the type of a Symbol \
+                           object class or one of its subclasses.
+        :raises TypeError: if the name argument is not a str.
+        '''
+        if not (isinstance(symbol_type, type) and
+                Symbol in inspect.getmro(symbol_type)):
+            raise TypeError(f"The symbol_type parameter should be a type class of "
+                            f"a sublcass of Symbol, but found {type(symbol_type)} instead.")
+        if not isinstance(name, str):
+            raise TypeError(f"The name parameter should be a str, but found {type(name)}.")
+    
+        visibility = Symbol.Visibility.LOCAL
+        if self._is_global:
+            visibility = Symbol.Visibility.GLOBAL
+    
+        symbol = symbol_type(name=name, datatype=datatype, visibility=visibility)
+        self.add(symbol)
+
+        return symbol
+
+    def lookup(self, name: str) -> Union[Symbol, NoneType]:
+        '''
+        Search for a Symbol named name in the symbol table. Returns the symbol
+        if it exists, otherwise returns None.
+
+        :param name: Name of the symbol to lookup.
+        :type name: str
+
+        :returns: The symbol found in the symbol table.
+        :rtype: Symbol or None.
+
+        :raises TypeError: If the name argument is not a string.
+        '''
+        if not isinstance(name, str):
+            raise TypeError("Expected the name argument to be a string, but got "
+                            f"{type(name)}.")
+
+        # TODO Should we look in the global symbol table if we don't find locally?
```

Comment:
This TODO isn't implemented but is probably needed now that the main DSL has support for global variables.